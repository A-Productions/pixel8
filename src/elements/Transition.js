import Pixel8Element from './Pixel8Element'
import easing from '../easing'

export default class Transition extends Pixel8Element {
  static defaultProps = {
    frames: 8,
    delay: 0,
    ease: x => x,
    use: (from, to, t) => {
      const next = { ...to }
      // default transition props whitelist
      const keys = ['x', 'y', 'width', 'height', 'radius', 'borderRadius']
      for (const k of keys) {
        const hasProp = k in to
        const hasDiff = from[k] !== to[k]
        if (!hasProp || !hasDiff) continue
        next[k] = from[k] + Math.round((to[k] - from[k]) * t)
      }
      return next
    },
  }
  init() {
    this.setState({
      tick: 0,
      wait: this.props.delay,
      next: new WeakMap(),
      prev: new WeakMap(),
    })
  }
  draw(f = x => x) {
    const { next } = this.state
    // only draw children
    for (const child of this.children) {
      child.draw(x => {
        const props = next.get(child) || x
        return f(props)
      })
    }
  }
  update() {
    const { state } = this
    for (const child of this.children) {
      const { props } = child
      const prev = state.prev.get(child) || props
      const next = state.next.get(child) || props
      const same = Object.keys(next).every(k => next[k] === props[k])
      if (same) {
        if (state.tick !== 0 && this.props.onTransitionEnd) {
          this.props.onTransitionEnd()
        }
        // already transitioned
        this.setState({
          tick: 0,
          wait: this.props.delay,
        })
        state.prev.set(child, props)
        state.next.set(child, props)
      } else if (state.wait) {
        // handle delay
        state.wait--
      } else {
        // transition!
        state.tick++
        const { ease, frames, use } = this.props
        const f = 'function' === typeof ease ? ease : easing[ease]
        // linear progress
        const t0 = state.tick / (frames - 1)
        // eased progress
        const t1 = f(t0)
        // calculate transitions
        const nextProps = use(prev, props, t1)
        // set next props
        state.next.set(child, nextProps)
      }
      // update child
      child.update()
    }
  }
}
