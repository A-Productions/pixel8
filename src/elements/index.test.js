import Rectangle from './Rectangle'
import Transition from './Transition'

class ImageData {
  constructor(w, h) {
    this.width = w
    this.height = h
    this.data = new Uint8ClampedArray(new ArrayBuffer(w * h * 4))
  }
}

const createRoot = (width, height) => {
  const imageData = new ImageData(width, height)
  const pixels = new Uint32Array(imageData.data.buffer)
  const hitmap = new Uint32Array(new ArrayBuffer(imageData.data.length))
  const children = new Set()
  const childMap = new Map()
  const registerChild = () => {}
  const unregisterChild = () => {}
  const appendChild = child => {
    children.add(child)
    childMap.set(child.id, child)
  }
  const removeChild = child => {
    children.delete(child)
    childMap.delete(child.id)
  }
  const draw = () => {
    for (let i = 0; i < pixels.length; i++) {
      pixels[i] = hitmap[i] = 0
    }
    for (const child of children) {
      child.draw()
    }
  }
  const getPixel = (x, y) => pixels[y * width + x]
  const getChild = (x, y) => childMap.get(hitmap[y * width + x])
  return {
    imageData,
    pixels,
    hitmap,
    children,
    childMap,
    appendChild,
    removeChild,
    registerChild,
    unregisterChild,
    draw,
    getPixel,
    getChild,
  }
}

const pixelsToString = root => {
  let px = []
  let i = 0
  for (let y = 0; y < root.imageData.height; y++) {
    px.push([])
    for (let x = 0; x < root.imageData.width; x++) {
      px[y].push(root.pixels[i])
      i++
    }
  }
  return '\n' + px.map(x => x.join(' ')).join('\n')
}

it('[0, 0] 1x1', () => {
  const props = {
    x: 0,
    y: 0,
    width: 1,
    height: 1,
    fill: 1,
  }
  const children = null
  const root = createRoot(8, 8)
  const rect = new Rectangle(props, null, root)
  root.appendChild(rect)
  root.draw()
  expect(root.getPixel(0, 0)).toBe(1)
  expect(root.getChild(0, 0)).toBe(rect)
  expect(pixelsToString(root)).toMatchSnapshot()
})

it('elements nested in transition', () => {
  const props = {
    x: 0,
    y: 0,
    width: 1,
    height: 1,
    fill: 1,
  }
  const children = null
  const root = createRoot(8, 8)
  const tran = new Transition(
    {
      frames: 4,
    },
    null,
    root,
  )
  const rect1 = new Rectangle(
    {
      x: 0,
      y: 0,
      width: 4,
      height: 4,
      fill: 1,
    },
    null,
    root,
  )
  const rect2 = new Rectangle(
    {
      x: 0,
      y: 0,
      width: 2,
      height: 2,
      fill: 2,
    },
    null,
    root,
  )
  root.appendChild(tran)
  tran.appendChild(rect1)
  rect1.appendChild(rect2)
  root.draw()
  expect(pixelsToString(root)).toMatchSnapshot()
  rect1.setProps({ x: 4 })
  for (let i = 0; i < 4; i++) {
    root.draw()
    expect(pixelsToString(root)).toMatchSnapshot()
  }
})
