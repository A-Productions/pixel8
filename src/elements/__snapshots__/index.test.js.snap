// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[0, 0] 1x1 1`] = `
"
1 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0"
`;

exports[`elements nested in transition 1`] = `
"
2 2 1 1 0 0 0 0
2 2 1 1 0 0 0 0
1 1 1 1 0 0 0 0
1 1 1 1 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0"
`;

exports[`elements nested in transition 2`] = `
"
0 2 2 1 1 0 0 0
0 2 2 1 1 0 0 0
0 1 1 1 1 0 0 0
0 1 1 1 1 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0"
`;

exports[`elements nested in transition 3`] = `
"
0 0 2 2 1 1 0 0
0 0 2 2 1 1 0 0
0 0 1 1 1 1 0 0
0 0 1 1 1 1 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0"
`;

exports[`elements nested in transition 4`] = `
"
0 0 0 2 2 1 1 0
0 0 0 2 2 1 1 0
0 0 0 1 1 1 1 0
0 0 0 1 1 1 1 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0"
`;

exports[`elements nested in transition 5`] = `
"
0 0 0 0 2 2 1 1
0 0 0 0 2 2 1 1
0 0 0 0 1 1 1 1
0 0 0 0 1 1 1 1
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0"
`;
